content/post/083-js-ipfs-0.41.md

```diff
@@ -0,0 +1,313 @@
+
+🚨 Oh, wait, serious note - this release brings **big breaking changes** to the core API so please consult the "[API Changes](#api-changes)" section below for all the information.
+
+It's been a long and emotional voyage but this refactor brings a plethora of incredible immediate and future benefits to consumers, contributors and core developers that make all the work and effort worthwhile. To summarise, we've:
+
+* Switched to streaming APIs by default to reduce memory pressure
+* Reduced API surface area by removing buffering, Node.js and Pull Stream APIs
+* Reduced the amount of code in the code base and number of dependencies we depend on (for smaller browser bundles and faster install times)
+* Switched to using async iterables to stream data to help make streaming more approachable
+* Switched to using `async`/`await` so we’ll get better error stack traces and improved readability and maintainability
+
+Using `async`/`await` in JavaScript is gaining a _lot_ of traction in the ecosystem and is rapidly becoming the de facto way of writing idiomatic JS. We want js-ipfs to **move with the times and continue to be attractive to contributors** by using modern JS features, techniques and practices. The big idea with these changes is for the code to be easier to contribute to, easier to understand, easier to maintain, and be faster and smaller than ever.
+
+This change is so big, and so significant, we wrote a whole [blog post](https://blog.ipfs.io/2020-02-01-async-await-refactor/) about it to explain the motivations behind the changes...in ~~excruciating~~ delightful depth 🤣!
+
+We've also compiled some stats on this refactor for your viewing pleasure:
+
+* 27 direct dependencies were removed from our `package.json`
+* 214 fewer modules end up in our browser bundle
+* 155 kB lighter browser bundle ([unpkg.com/ipfs@0.40.0](https://unpkg.com/ipfs@0.40.0/dist/index.min.js) vs [unpkg.com/ipfs@0.41.0-rc.0](https://unpkg.com/ipfs@0.41.0-rc.0/dist/index.min.js)) - that's **18% smaller**!
+* 124 kB lighter `ipfs-http-client` browser bundle ([unpkg.com/ipfs-http-client@41.0.1](https://unpkg.com/ipfs-http-client@41.0.1/dist/index.min.js) vs [unpkg.com/ipfs-http-client@42.0.0](https://unpkg.com/ipfs-http-client@42.0.0/dist/index.min.js)) - that's **60% smaller**!
+* ~2,600 lines of code removed (net)
+* ~360 lines of code removed from `ipfs-http-client` (net)
+* ~10 minutes shorter CI run times
+
+...and a lot of those stats are just for js-ipfs and js-ipfs-http-client - the tip of the iceberg! We saw changes similar to this for [between 60-70 dependencies](https://github.com/ipfs/js-ipfs/issues/1670) across IPFS, libp2p, IPLD and multiformats.
+
+## 🌗 UnixFS v1.5
```

Comment:
Do you think it's worth talking about `mode` a bit more?  The idea was that package managers need to know which files are executable and by who, and the combination of that, `mtime` and mounting your mfs with fuse would give you an actual functioning unix file system backed by IPFS.

Reply:
Though the mount work sort of fell by the wayside..

Reply:
It's just...not that interesting 😂

Is there more to say than package managers need to know that info to be able to be a package manager? With `mtime` there's a real problem that people can understand and see how it would lead to a performance gain.

Reply:
It's up to you, but being able to persist the file mode opens up NFS type use cases that weren't possible before, like having your `node_modules` folder be on IPFS and everything under `node_modules/.bin` actually being executable.

Reply:
Thanks, that's interesting! I added some words about this https://github.com/ipfs/blog/pull/369/commits/3bc5283f65f94589074ecdc626bfd802ffba927d (mostly your words 😉)